lte/gateway/c/oai/tasks/amf/amf_app_ue_context_and_proc.h

```diff
@@ -0,0 +1,1001 @@
+typedef int (*pdu_in_resp_t)(
+    amf_context_t* amf_ctx, void* arg);  // can be RESPONSE, COMPLETE, ACCEPT
+typedef int (*pdu_in_rej_t)(amf_context_t* amf_ctx, void* arg);  // REJECT.
+typedef int (*pdu_out_rej_t)(
+    amf_context_t* amf_ctx, nas5g_base_proc_t* nas_proc);  // REJECT.
+typedef void (*time_out_t)(void* arg);
+typedef int (*sdu_out_delivered_t)(
+    amf_context_t* amf_ctx, nas_amf_proc_t* nas_proc);
+typedef int (*sdu_out_not_delivered_t)(
+    amf_context_t* amf_ctx, nas_amf_proc_t* nas_proc);
+typedef int (*sdu_out_not_delivered_ho_t)(
+    amf_context_t* amf_ctx, nas_amf_proc_t* nas_proc);
+
+/*Global variable and needed to increment based on nas procedures*/
+static uint64_t nas_puid = 1;
+
+class nas5g_base_proc_t {
+ public:
+  success_cb_t success_notif;
+  failure_cb_t failure_notif;
+  proc_abort_t abort;
+  pdu_in_rej_t fail_in;
+  pdu_out_rej_t fail_out;
+  time_out_t time_out;
+  nas_base_proc_type_t type;  // AMF, SMF, CN
+  uint64_t nas_puid;          // procedure unique identifier for internal use
+
+  class nas5g_base_proc_t* parent;
+  class nas5g_base_proc_t* child;
+};
+
+enum nas_amf_proc_type_t {
+  NAS_AMF_PROC_TYPE_NONE = 0,
+  NAS_AMF_PROC_TYPE_SPECIFIC,
+  NAS_AMF_PROC_TYPE_COMMON,
+  NAS_AMF_PROC_TYPE_CONN_MNGT,
+};
+
+// AMF Specific procedures
+class nas_amf_proc_t {
+ public:
+  nas5g_base_proc_t base_proc;
+  nas_amf_proc_type_t type;  // specific, common, connection management
+
+  // SDU interface
+  sdu_out_delivered_t delivered;
+  sdu_out_not_delivered_t not_delivered;
+  sdu_out_not_delivered_ho_t not_delivered_ho;
+  amf_fsm_state_t previous_amf_fsm_state;
+};
+
+typedef struct nas_amf_common_proc_s {
+  nas_amf_proc_t amf_proc;
+  amf_common_proc_type_t type;
+} nas_amf_common_proc_t;
+
+enum amf_specific_proc_type_t {
+  AMF_SPEC_PROC_TYPE_NONE = 0,
+  AMF_SPEC_PROC_TYPE_REGISTRATION,
+  AMF_SPEC_PROC_TYPE_DEREGISTRATION,
+  AMF_SPEC_PROC_TYPE_TAU,
+};
+
+/*Deregistration specific elements*/
+typedef enum deregistration_switchoff_e {
+  /*In the UE to network direction, octate 1, 4th bit*/
+  AMF_NORMAL_DEREGISTRATION = 0,
+  AMF_SWITCHOFF_DEREGISTRATION,
+} deregistration_switchoff;
+
+typedef enum deregistration_access_type_e {
+  AMF_3GPP_ACCESS = 1,
+  AMF_NONE_3GPP_ACCESS,
+  AMF_3GPP_ACCESS_AND_NONE_3GPP_ACCESS,
+} deregistration_access_type;
+
+typedef struct amf_deregistration_request_ies_s {
+  deregistration_switchoff de_reg_type;
+  deregistration_access_type de_reg_access_type;
+  ksi_t ksi;
+  TmsiM5GSMobileIdentity* tmsi;
+  SuciM5GSMobileIdentity* suci;
+  ImsiM5GSMobileIdentity* imsi;
+  ImeiM5GSMobileIdentity* imei;
+  GutiM5GSMobileIdentity* guti;
+  amf_nas_message_decode_status_t decode_status;
+} amf_deregistration_request_ies_t;
+
+// AMF Specific procedures
+class nas_amf_specific_proc_t {
+ public:
+  nas_amf_proc_t amf_proc;
+  amf_specific_proc_type_t type;
+};
+
+class identification : public amf_context_t {
```

Comment:
this class definition, particularly the inheritance does not make much sense.

Reply:
fixed